source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -607,7 +658,23 @@ Otherwise, the `Connection <#connection>`_ is marked as available.
 Clearing a Connection Pool
 --------------------------
 
-A Pool MUST have a method of clearing all `Connections <#connection>`_ when instructed. Rather than iterating through every `Connection <#connection>`_, this method should simply increment the generation of the Pool, implicitly marking all current `Connections <#connection>`_ as stale. The checkOut and checkIn algorithms will handle clearing out stale `Connections <#connection>`_. If a user is subscribed to Connection Monitoring events, a PoolClearedEvent MUST be emitted after incrementing the generation.
+A Pool MUST have a method of clearing all `Connections <#connection>`_ when
+instructed. Rather than iterating through every `Connection <#connection>`_,
+this method should simply increment the generation of the Pool, implicitly
+marking all current `Connections <#connection>`_ as stale. It should also
+transition the pool's state to "paused" to halt the creation of new connections
+until it is marked as "ready" again. The checkOut and checkIn algorithms will
+handle clearing out stale `Connections <#connection>`_. If a user is subscribed
+to Connection Monitoring events, a PoolClearedEvent MUST be emitted after
+incrementing the generation / marking the pool as "paused".
+
+As part of clearing the pool, the WaitQueue MUST also be cleared, meaning all
+requests in the WaitQueue MUST fail with errors indicating that the pool was
+cleared while the checkOut was being performed. The error returned MUST be
+considered a retryable error. Clearing the WaitQueue MUST happen eagerly so that
+any operations waiting on `Connections <#connection>`_ can retry as soon as
+possible. The pool MUST NOT rely on WaitQueueTimeoutMS to clear requests from
+the WaitQueue.
```

Comment:
Wouldn't this be the first driver made retryable error? I assume it should follow the same rules as other retryable errors, so would retryReads=false make this throw?

Reply:
Drivers already attach `RetryableWriteError` label to network errors (which this basically is). See here: https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#determining-retryable-errors

And yeah, in the case of retryWrites=false, this would fail those operations. Note that these operations would likely fail anyways if retryWrites=false, just that it happens sooner now and without attempting a connection creation.